service/controller/clusterapi/v16/resources/awsclusterconfig/create.go

```diff
@@ -21,7 +26,125 @@ func (r *Resource) EnsureCreated(ctx context.Context, obj interface{}) error {
+			return nil
+		} else if err != nil {
+			return microerror.Mask(err)
+		}
+
+		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("created AWSClusterConfig %q/%q", awsClusterConfig.Namespace, awsClusterConfig.Name))
+	}
+
+	return nil
+}
+
+// getAWSClusterConfig returns corresponding AWSClusterConfig CR for given
+// Cluster if one exists or constructs new empty one.
+func (r *Resource) getAWSClusterConfig(ctx context.Context, cluster clusterv1alpha1.Cluster) (*v1alpha1.AWSClusterConfig, error) {
+	var awsClusterConfig *v1alpha1.AWSClusterConfig
+	var err error
+
+	awsClusterConfigName := key.AWSClusterConfigName(cluster)
+
+	r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("finding out if AWSClusterConfig %q/%q exists", cluster.Namespace, awsClusterConfigName))
+
+	awsClusterConfig, err = r.g8sClient.CoreV1alpha1().AWSClusterConfigs(cluster.Namespace).Get(awsClusterConfigName, v1.GetOptions{})
+	if errors.IsNotFound(err) {
+		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("did not find AWSClusterConfig %q/%q", cluster.Namespace, awsClusterConfigName))
+		awsClusterConfig = constructAWSClusterConfig(cluster)
+	} else if err != nil {
+		return nil, microerror.Mask(err)
+	} else {
+		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("found AWSClusterConfig %q/%q", cluster.Namespace, awsClusterConfigName))
+	}
+
+	return awsClusterConfig, nil
+}
+
+func (r *Resource) getMachineDeployments(ctx context.Context, cluster clusterv1alpha1.Cluster) ([]clusterv1alpha1.MachineDeployment, error) {
+	labelSelector := v1.AddLabelToSelector(&v1.LabelSelector{}, label.Cluster, key.ClusterID(cluster))
+	// TODO: Add selector for provider annotation?
+
+	listOptions := v1.ListOptions{
+		LabelSelector: labelSelector.String(),
+	}
+
+	machineDeploymentList, err := r.cmaClient.ClusterV1alpha1().MachineDeployments(cluster.Namespace).List(listOptions)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	return machineDeploymentList.Items, nil
+}
+
+func (r *Resource) mapClusterToAWSClusterConfig(awsClusterConfig *v1alpha1.AWSClusterConfig, cluster clusterv1alpha1.Cluster, machineDeployments []clusterv1alpha1.MachineDeployment) error {
+	awsClusterConfig.Spec.Guest.ClusterGuestConfig.AvailabilityZones = len(key.ClusterAvailabilityZones(cluster, machineDeployments))
+	awsClusterConfig.Spec.Guest.ClusterGuestConfig.DNSZone = key.ClusterDNSZone(cluster)
+	awsClusterConfig.Spec.Guest.ClusterGuestConfig.ID = key.ClusterID(cluster)
+	awsClusterConfig.Spec.Guest.ClusterGuestConfig.Name = key.ClusterName(cluster)
+	awsClusterConfig.Spec.Guest.ClusterGuestConfig.ReleaseVersion = key.ClusterReleaseVersion(cluster)
+
+	// Are these needed?
+	// awsClusterConfig.Spec.Guest.ClusterGuestConfig.Owner
+	// awsClusterConfig.Spec.Guest.ClusterGuestConfig.VersionBundles
```

Comment:
I think the version bundle is because we still do versioning and controller routing with it but the owner is not needed as we also start to derive it from some label.

Reply:
I thought about this and my current plan is to copy release lister implementation from `api` to `service/release` in a separate PR and use it here to fetch version bundles for given release version. Would that be ok?

Reply:
That is the only option we have anyway I guess. And it is kind of temporary. So let's do it that way. 👍